### PR TITLE
SI-8146 Fix non-deterministic <:< for deeply nested types

### PR DESCRIPTION
**Pull request by retronym**
Details in the commit messages.

Review by @adriaanm, @odersky /cc @szeiger

### CODE REVIEW

File: src/reflect/scala/reflect/internal/tpe/TypeComparers.scala [hunk 1]
```diff
@@ -17,20 +17,15 @@ trait TypeComparers {
...
-        val tvars = List(tp1, stp.tp1, tp2, stp.tp2) flatMap (t => if (t.isGround) Nil else typeVarsInType(t))
-        suspendingTypeVars(tvars)(tp1 =:= stp.tp1 && tp2 =:= stp.tp2)
-      case _ =>
-        false
-    })
+  final case class SubTypePair(tp1: Type, tp2: Type) {
+    // SI-8146 we used to implement equality here in terms of pairwise =:=.
+    //         But, this was inconsistent with hashCode, which was based on the
+    //         Type#hashCode, based on the structure of types, not the meaning.
+    //         Now, we use `Type#{equals,hashCode}` as the (consistent) basis for
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:24]: How did this lead to non-determinism previously?

Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:24]: If `equals` and `hashCode` don't follow the contract, and at some-point `Object#hashCode` contributes to your hash code (e.g. TypeVar, see 7e52fb9), and you use `Hash{Set,Map}` (here: `pendingSubtypes`), things become non-deterministic, with JVM object IDs as the source of entropy.

**Comment by retronym** [comment]
We might also consider this change for 2.10.x. It was reported in the context of Slick. Reviewers, WDTY?

**Comment by adriaanm** [comment]
LGTM!

**Comment by adriaanm** [comment]
Safe to backport, I'd say. I can't think of a way this would change type inference (before we suspended them and called =:=, now we go straight to reference-equality). That is, assuming we actually didn't miss any TypeVars that needed to be suspended. (One way that would've happened is for TypeVars with a TypeVar as their `origin`.)

**Comment by retronym** [comment]
Backport: #3367

**Pull request merged**